features/285-design-webhooks/design.md

```diff
@@ -0,0 +1,129 @@
+# next webhooks design
+
+## Overview
+
+skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
+
+When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+
+## SYNC and ASYNC hooks
+
+There are two types of webhook:
+
+- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
+- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+
+## Request Headers
+
+Following headers will be added to webhook POST request:
+
+- `X-Skygear-Webhook-Req-ID`: an ID of the request.
+- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
+- `X-Skygear-Webhhok-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+
+## Request Payload
+
+Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+
+All event payloads follows following payload format:
+
+```json=
+{
+    "event": "event_name",
+    "data": {
+        "key_name_1": "event_data_key_value",
+        "key_name_2": "event_data_key_value",
+        ...,
+        "context": {
+            "user": <current user object>,
+            "req": {
+                "path": "request_path",
+                "body": <original_request_body>,
+                "id": "request_id"
+            }
+        }
+    }
+}
+```
+
+Note that, `context` is above format is still under discussion.
+
+## Responding to a webhook
+
+To acknowledge receipt of a webhook, webhook endpoint should return a 2xx HTTP status code. All response codes outside this range, including 3xx codes, will indicate to skygear the request is failed.
+
+For some events, it may allow to webhook to modify changes (e.x. auth gear's before family hooks) when response code is 2xx. The behavior is defined by each gear, so will not be covered in this spec.
+
+Take auth gear as an example, its before family hooks accept that a hook returns a user object to modify current user object.
```

Comment:
Do we support accepting empty body if the hook does not need to modify the object? Or we would simply suggest the developer return the original object.

Reply:
I think missing fields means unmodified, so empty body is accepted. Another thing is gear should validate return object, if failed, gear should return an error to SDK. Do you think it is okay?

Reply:
Got it.